### PR TITLE
[Enhancement](plan) Optimize preagg for aggregate function

### LINKED ISSUE #8562: [Bug] 特定过滤无法命中物化视图
### Search before asking

- [X] I had searched in the [issues](https://github.com/apache/incubator-doris/issues?q=is%3Aissue) and found no similar issues.

### Version

0.14.13

### What's Wrong?

create table sql：

CREATE TABLE `core_doris_driver_cargo_action_funnel_di_driver`(
  `driver_id` largeint COMMENT '司机id', 
  `is_click` largeint COMMENT '是否点击', 
  `is_call` int COMMENT '是否有拨打行为', 
  `is_fd` int COMMENT '是否有找货行为', 
  `is_deal` int COMMENT '是否成交', 
  `is_zq` int COMMENT '是否直抢', 
  `day` int COMMENT '发货日期',
  `month` int(11) NULL COMMENT "月份",
  `count` int(11) SUM NULL COMMENT "发货量"
  )ENGINE=OLAP
AGGREGATE KEY(`driver_id`,`is_click`,`is_call`,`is_fd`,`is_deal`,`is_zq`,`day`,`month`)
COMMENT "司机行为"
PARTITION BY RANGE(day)
(
    PARTITION p20191231 VALUES LESS THAN ('20200101'),
    PARTITION p20201231 VALUES LESS THAN ('20210101'),
    PARTITION p20210630 VALUES LESS THAN ('20210701'),
	PARTITION p20210731 VALUES LESS THAN ('20210801'),
	PARTITION p20210831 VALUES LESS THAN ('20210901'),
	PARTITION p20210901 VALUES LESS THAN ('20210902'),
	PARTITION p20210902 VALUES LESS THAN ('20210906'),
	PARTITION p20210906 VALUES LESS THAN ('20210907'),
	PARTITION p20211015 VALUES LESS THAN ('20211016'),
	PARTITION p20211016 VALUES LESS THAN ('20211017'),
	PARTITION p20211017 VALUES LESS THAN ('20211018'),
	PARTITION p20211018 VALUES LESS THAN ('20211019'),
	PARTITION p20211019 VALUES LESS THAN ('20211020'),
	PARTITION p20211020 VALUES LESS THAN ('20211021'),
	PARTITION p20211021 VALUES LESS THAN ('20211022'),
	PARTITION p20211031 VALUES LESS THAN ('20211101'),
	PARTITION p20211101 VALUES LESS THAN ('20211102'),
	PARTITION p20211102 VALUES LESS THAN ('20211103'),
	PARTITION p20211103 VALUES LESS THAN ('20211104'),
	PARTITION p20211231 VALUES LESS THAN ('20220101'),
	PARTITION p20220110 VALUES LESS THAN ('20220111'),
	PARTITION p20220111 VALUES LESS THAN ('20220112'),
	PARTITION p20220112 VALUES LESS THAN ('20220113'),
	PARTITION p20220113 VALUES LESS THAN ('20220114'),
	PARTITION p20220114 VALUES LESS THAN ('20220115'),
	PARTITION p20220125 VALUES LESS THAN ('20220126'),
	PARTITION p20220126 VALUES LESS THAN ('20220127'),
	PARTITION p20220127 VALUES LESS THAN ('20220128'),
	PARTITION p20220128 VALUES LESS THAN ('20220129')
)
DISTRIBUTED BY HASH(`driver_id`) BUCKETS 12
PROPERTIES (
"replication_num" = "1",
"in_memory" = "false",
"storage_format" = "V2"
);

insert sql : 
insert into core_doris_driver_cargo_action_funnel_di_driver values(1,1,1,1,1,1,20220101,202201,1)

MATERIALIZED VIEW sql:
  create materialized view funnel_driver_day_is_fd as select day, driver_id, is_fd,
  sum(count)
 from core_doris_driver_cargo_action_funnel_di_driver group by day, driver_id, is_fd;

query sql:
select sum(if(is_fd =1, count, null)) from core_doris_driver_cargo_action_funnel_di_driver
where day between 20220101 and 20220201 不能命中物化视图

select sum(count) from core_doris_driver_cargo_action_funnel_di_driver
where day between 20220101 and 20220201 and is_fd = 1 可以命中物化视图

 

### What You Expected?

select sum(count) from core_doris_driver_cargo_action_funnel_di_driver
where day between 20220101 and 20220201 and is_fd = 1 可以命中物化视图

### How to Reproduce?

_No response_

### Anything Else?

_No response_

### Are you willing to submit PR?

- [ ] Yes I am willing to submit a PR!

### Code of Conduct

- [X] I agree to follow this project's [Code of Conduct](https://www.apache.org/foundation/policies/conduct)

### PR DESCRIPTION
**Pull request by DarvenDuan**
If the `case when` contains a constant zero/null in SUM agg function, we can use preaggregation to speed up queries and it will not effect the correctness of result.

Issue Number: close #8562

If we use `case when` or `if` in `SUM`, Planner will get the follows steps:
1、rewrite expression `if` to `case when`.
2、check the Index contains all key columns in `when` expr.
3、check the expr in `then` is constant zero、null or value column in index.
4、check the expr in `else` is constant zero、null or value column in Index.

If conditions 2, 3, and 4 are all met, we open preagg.

## Further comments

If this is a relatively large or complex change, kick off the discussion at [dev@doris.apache.org](mailto:dev@doris.apache.org) by explaining why you chose the solution you did and what alternatives you considered, etc...

### CODE REVIEW

**Comment by DarvenDuan** [comment]
run p0

**Comment by DarvenDuan** [comment]
run compile

**Comment by DarvenDuan** [comment]
run compile

**Comment by DarvenDuan** [comment]
run p0

**Comment by DarvenDuan** [comment]
run p1

**Comment by DarvenDuan** [comment]
run fe-ut

**Comment by DarvenDuan** [comment]
run be-ut

**Comment by DarvenDuan** [comment]
run fe_ut

**Comment by DarvenDuan** [comment]
run buildall

**Comment by DarvenDuan** [comment]
run p0

**Comment by DarvenDuan** [comment]
run p0

**Comment by DarvenDuan** [comment]
run p0

**Comment by DarvenDuan** [comment]
run compile

**Comment by DarvenDuan** [comment]
run external

**Comment by DarvenDuan** [comment]
run pipline_x

**Comment by DarvenDuan** [comment]
run pipeline_x

**Comment by DarvenDuan** [comment]
run pipeline_x p0

**Comment by DarvenDuan** [comment]
run pipelinex_p0

**Comment by DarvenDuan** [comment]
run feut

**Comment by DarvenDuan** [comment]
run beut

**Comment by BiteTheDDDDt** [comment]
I don't understand why we need to support this case, it seems meaningless?

**Comment by DarvenDuan** [comment]
> I don't understand why we need to support this case, it seems meaningless?

If we create a rollup or a materialized view for  data preaggregation, and then use `SUM(case ... when ... then)` or `SUM(if(... , ... , ...))`, Doris can't hit rollup or materialized view we created before.

simple case:

create table:
CREATE TABLE test_tbl (
  key1 INT NOT NULL,
  key2 INT NULL,
  key3 INT NULL,
  value BIGINT sum NULL
) ENGINE=OLAP
AGGREGATE KEY(key1, key2, key3)
COMMENT 'OLAP'
DISTRIBUTED BY HASH(key1) BUCKETS 10
PROPERTIES (
"replication_num" = "1"
);

create rollup:
`alter table test_tbl add rollup test_rollup(key1,key2,value);`

I hope this query can hit rollup **test_rollup**
`select key1,sum(case when key2 > 0 then value else 0 end) as value from test_tbl group by key1;`

**Comment by BiteTheDDDDt** [comment]
> > I don't understand why we need to support this case, it seems meaningless?
> 
> If we create a rollup or a materialized view for data preaggregation, and then use `SUM(case ... when ... then)` or `SUM(if(... , ... , ...))`, Doris can't hit rollup or materialized view we created before.
> 
> simple case:
> 
> create table: CREATE TABLE test_tbl ( key1 INT NOT NULL, key2 INT NULL, key3 INT NULL, value BIGINT sum NULL ) ENGINE=OLAP AGGREGATE KEY(key1, key2, key3) COMMENT 'OLAP' DISTRIBUTED BY HASH(key1) BUCKETS 10 PROPERTIES ( "replication_num" = "1" );
> 
> create rollup: `alter table test_tbl add rollup test_rollup(key1,key2,value);`
> 
> I hope this query can hit rollup **test_rollup** `select key1,sum(case when key2 > 0 then value else 0 end) as value from test_tbl group by key1;`

Can this kind of rewriting be used as a general rule instead of targeting materialized views?

**Comment by DarvenDuan** [comment]
> > > I don't understand why we need to support this case, it seems meaningless?
> > 
> > 
> > If we create a rollup or a materialized view for data preaggregation, and then use `SUM(case ... when ... then)` or `SUM(if(... , ... , ...))`, Doris can't hit rollup or materialized view we created before.
> > simple case:
> > create table: CREATE TABLE test_tbl ( key1 INT NOT NULL, key2 INT NULL, key3 INT NULL, value BIGINT sum NULL ) ENGINE=OLAP AGGREGATE KEY(key1, key2, key3) COMMENT 'OLAP' DISTRIBUTED BY HASH(key1) BUCKETS 10 PROPERTIES ( "replication_num" = "1" );
> > create rollup: `alter table test_tbl add rollup test_rollup(key1,key2,value);`
> > I hope this query can hit rollup **test_rollup** `select key1,sum(case when key2 > 0 then value else 0 end) as value from test_tbl group by key1;`
> 
> Can this kind of rewriting be used as a general rule instead of targeting materialized views?

It not just works on Materialized views. I added this rewriting rule in both `turnOffPreAgg ` function for legacy planner and `checkAggFunc` function for nereids planner. So it can work as a general rule.

**Comment by starocean999** [comment]
> > I don't understand why we need to support this case, it seems meaningless?
> 
> If we create a rollup or a materialized view for data preaggregation, and then use `SUM(case ... when ... then)` or `SUM(if(... , ... , ...))`, Doris can't hit rollup or materialized view we created before.
> 
> simple case:
> 
> create table: CREATE TABLE test_tbl ( key1 INT NOT NULL, key2 INT NULL, key3 INT NULL, value BIGINT sum NULL ) ENGINE=OLAP AGGREGATE KEY(key1, key2, key3) COMMENT 'OLAP' DISTRIBUTED BY HASH(key1) BUCKETS 10 PROPERTIES ( "replication_num" = "1" );
> 
> create rollup: `alter table test_tbl add rollup test_rollup(key1,key2,value);`
> 
> I hope this query can hit rollup **test_rollup** `select key1,sum(case when key2 > 0 then value else 0 end) as value from test_tbl group by key1;`

Thanks for your effort. But this pr is not correct, because the sql above SHOULD NOT hit mv. The sql bellow is correct one which need to hit mv(but not supported now). 
`select key1,sum(case when key2 > 0 then value else 0 end) as value from test_tbl where key2 > 0 group by key1;`
This sql is equivalent to  
`select key1,sum(value) as value from test_tbl where key2 > 0 group by key1;`

As the mv selection logical is being refactoring in nereids. We can considering such optimization after the refactoring work done. ( This refactoring is scheduled, but not in near future)

**Comment by DarvenDuan** [comment]
> > > I don't understand why we need to support this case, it seems meaningless?
> > 
> > 
> > If we create a rollup or a materialized view for data preaggregation, and then use `SUM(case ... when ... then)` or `SUM(if(... , ... , ...))`, Doris can't hit rollup or materialized view we created before.
> > simple case:
> > create table: CREATE TABLE test_tbl ( key1 INT NOT NULL, key2 INT NULL, key3 INT NULL, value BIGINT sum NULL ) ENGINE=OLAP AGGREGATE KEY(key1, key2, key3) COMMENT 'OLAP' DISTRIBUTED BY HASH(key1) BUCKETS 10 PROPERTIES ( "replication_num" = "1" );
> > create rollup: `alter table test_tbl add rollup test_rollup(key1,key2,value);`
> > I hope this query can hit rollup **test_rollup** `select key1,sum(case when key2 > 0 then value else 0 end) as value from test_tbl group by key1;`
> 
> Thanks for your effort. But this pr is not correct, because the sql above SHOULD NOT hit mv. The sql bellow is correct one which need to hit mv(but not supported now). `select key1,sum(case when key2 > 0 then value else 0 end) as value from test_tbl where key2 > 0 group by key1;` This sql is equivalent to `select key1,sum(value) as value from test_tbl where key2 > 0 group by key1;`
> 
> As the mv selection logical is being refactoring in nereids. We can considering such optimization after the refactoring work done. ( This refactoring is scheduled, but not in near future)

Thanks for your replay, I have a few questions I'd like to ask you, can I add your WeChat?  My WeChat: DUAN1998525

**Comment by DarvenDuan** [comment]
run buildall

**Comment by DarvenDuan** [comment]
run feut

**Comment by DarvenDuan** [comment]
run buildall

**Comment by DarvenDuan** [comment]
run p0

**Comment by DarvenDuan** [comment]
run arm

File: fe/fe-core/src/main/java/org/apache/doris/nereids/trees/expressions/Expression.java [hunk 1]
```diff
@@ -260,6 +260,10 @@ public boolean isConstant() {
         }
     }
 
+    public boolean isConstantZero() {
```
Comment (starocean999) [inline fe/fe-core/src/main/java/org/apache/doris/nereids/trees/expressions/Expression.java:263]: using name isLiteralZero is better

File: fe/fe-core/src/test/java/org/apache/doris/nereids/rules/rewrite/mv/SelectMvIndexTest.java [hunk 2]
```diff
@@ -1251,4 +1251,88 @@ private void testMvWithTwoTable(String sql, String firstTableIndexName, String s
             Assertions.assertEquals(secondTableIndexName, scan1.getSelectedIndexName());
         });
     }
+
+    @Test
+    public void testSubExpressionsInAggregation() throws Exception {
+        createTable("CREATE TABLE db1.`test_pre_agg_tbl` (\n"
```
Comment (starocean999) [inline fe/fe-core/src/test/java/org/apache/doris/nereids/rules/rewrite/mv/SelectMvIndexTest.java:1257]: pls add test for float/double, decimalv2/v3 too

**Comment by DarvenDuan** [comment]
run buildall

**Comment by DarvenDuan** [comment]
run buildall

**Review by starocean999 — APPROVED** [review]
LGTM

**Review by BiteTheDDDDt — APPROVED** [review]

**Pull request merged**